mailpoet/lib/Segments/DynamicSegments/SegmentSaveController.php

```diff
@@ -37,4 +43,23 @@ public function save(array $data = []): SegmentEntity {
 
     return $this->segmentsRepository->createOrUpdate($name, $description, SegmentEntity::TYPE_DYNAMIC, $filtersData, $id);
   }
+
+  public function duplicate(SegmentEntity $segmentEntity): SegmentEntity {
+    $duplicate = clone $segmentEntity;
+    // translators: %s is the name of the segment
+    $duplicate->setName(sprintf(__('Copy of %s', 'mailpoet'), $segmentEntity->getName()));
+    $this->segmentsRepository->verifyNameIsUnique($duplicate->getName(), $duplicate->getId());
+    $this->entityManager->wrapInTransaction(function(EntityManager $entityManager) use ($duplicate, $segmentEntity) {
```

Comment:
Thanks! But I can't take credit for the idea. We do it in a few other places, including duplicating lists/non-dynamic segments (see [here](https://github.com/mailpoet/mailpoet/blob/38e4b4d2faf3a36271c0f70d330073ed1ebfb484/mailpoet/lib/Segments/SegmentSaveController.php#L51-L51)). Those other places are using the deprecated-as-of-doctrine-orm-2.10 `transactional` instead of `wrapInTransaction`.

I agree that more transactions could be a great idea for related operations like these. The only potential downside I could see to using transactions more frequently is that they're not supported by MyISAM. There are still plenty of people who have MyISAM tables, so we could end up with different behavior for different configurations.